File: source/c/simple_repl.c
// gcc -o simple_repl simple_repl.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <ctype.h>

typedef struct { char tokens[16][32]; int count; } Tokens;

Tokens tokenize(const char *expr) {
    Tokens t = {{}, 0};
    char num[32]; int ni = 0;
    for (int i = 0; expr[i]; i++) {
        char ch = expr[i];
        if (ch == ' ') {
            if (ni > 0) { num[ni] = '\0'; strcpy(t.tokens[t.count++], num); ni = 0; }
        } else if (ch == '+' || ch == '-' || ch == '*' || ch == '/') {
            if (ni > 0) { num[ni] = '\0'; strcpy(t.tokens[t.count++], num); ni = 0; }
            t.tokens[t.count][0] = ch; t.tokens[t.count][1] = '\0'; t.count++;
        } else {
            num[ni++] = ch;
        }
    }
    if (ni > 0) { num[ni] = '\0'; strcpy(t.tokens[t.count++], num); }
    return t;
}

double evaluate(const char *expr) {
    Tokens t = tokenize(expr);
    if (t.count == 0) return 0;
    double result = atof(t.tokens[0]);
    for (int i = 1; i + 1 < t.count; i += 2) {
        char op = t.tokens[i][0];
        double operand = atof(t.tokens[i+1]);
        if (op == '+') result += operand;
        else if (op == '-') result -= operand;
        else if (op == '*') result *= operand;
        else if (op == '/') result /= operand;
    }
    return result;
}

int main(void) {
    const char *inputs[] = {"2 + 3", "10 * 4", "100 / 7", "50 - 18", "2 + 3 * 4"};
    int ni = 5;
    printf("=== Simple Calculator ===\n");
    for (int i = 0; i < ni; i++) {
        double result = evaluate(inputs[i]);
        long long int_result = (long long)result;
        if ((double)int_result == result) printf("> %s = %lld\n", inputs[i], int_result);
        else printf("> %s = %g\n", inputs[i], result);
    }
    printf("=== Done ===\n");
    return 0;
}

File: source/go/simple_repl.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

func tokenize(expr string) []string {
	var tokens []string
	num := ""
	for i := 0; i < len(expr); i++ {
		ch := expr[i]
		if ch == ' ' {
			if num != "" {
				tokens = append(tokens, num)
				num = ""
			}
		} else if ch == '+' || ch == '-' || ch == '*' || ch == '/' {
			if num != "" {
				tokens = append(tokens, num)
				num = ""
			}
			tokens = append(tokens, string(ch))
		} else {
			num += string(ch)
		}
	}
	if num != "" {
		tokens = append(tokens, num)
	}
	return tokens
}

func evaluate(expr string) float64 {
	tokens := tokenize(expr)
	if len(tokens) == 0 {
		return 0
	}
	result, _ := strconv.ParseFloat(tokens[0], 64)
	i := 1
	for i < len(tokens) {
		op := tokens[i]
		operand, _ := strconv.ParseFloat(tokens[i+1], 64)
		switch op {
		case "+":
			result += operand
		case "-":
			result -= operand
		case "*":
			result *= operand
		case "/":
			result /= operand
		}
		i += 2
	}
	return result
}

func main() {
	inputs := []string{
		"2 + 3",
		"10 * 4",
		"100 / 7",
		"50 - 18",
		"2 + 3 * 4",
	}

	fmt.Println("=== Simple Calculator ===")
	for _, input := range inputs {
		result := evaluate(input)
		intResult := int64(result)
		if float64(intResult) == result {
			fmt.Printf("> %s = %d\n", input, intResult)
		} else {
			// Format like Hemlock: remove trailing zeros
			s := strconv.FormatFloat(result, 'f', -1, 64)
			if strings.Contains(s, ".") {
				s = strings.TrimRight(s, "0")
				s = strings.TrimRight(s, ".")
			}
			fmt.Printf("> %s = %s\n", input, s)
		}
	}
	fmt.Println("=== Done ===")
}
